pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java

```diff
@@ -284,4 +328,306 @@ public static TableRebalanceProgressStats.RebalanceStateStats getDifferenceBetwe
+    if (latestStepStats._totalCarryOverSegmentsToBeDeleted > 0) {
+      newOverallProgressStats._totalRemainingSegmentsToBeDeleted =
+          overallProgressStats._totalRemainingSegmentsToBeDeleted;
+    } else {
+      newOverallProgressStats._totalRemainingSegmentsToBeDeleted = numAdditionalSegmentsDeleted == 0
+          ? overallProgressStats._totalRemainingSegmentsToBeDeleted - numSegmentDeletesProcessedInLastStep
+          : overallProgressStats._totalRemainingSegmentsToBeDeleted + numSegmentDeletesProcessedInLastStep;
+    }
+    newOverallProgressStats._totalCarryOverSegmentsToBeDeleted =
+        latestStepStats._totalCarryOverSegmentsToBeDeleted;
+    newOverallProgressStats._totalRemainingSegmentsToConverge = latestStepStats._totalRemainingSegmentsToConverge;
+    newOverallProgressStats._totalUniqueNewUntrackedSegmentsDuringRebalance =
+        overallProgressStats._totalUniqueNewUntrackedSegmentsDuringRebalance + numberNewUntrackedSegmentsAdded;
+    newOverallProgressStats._percentageTotalSegmentsAddsRemaining =
+        calculatePercentageChange(newOverallProgressStats._totalSegmentsToBeAdded,
+            newOverallProgressStats._totalRemainingSegmentsToBeAdded
+                + newOverallProgressStats._totalCarryOverSegmentsToBeAdded);
+    newOverallProgressStats._percentageTotalSegmentDeletesRemaining =
+        calculatePercentageChange(newOverallProgressStats._totalSegmentsToBeDeleted,
+            newOverallProgressStats._totalRemainingSegmentsToBeDeleted
+                + newOverallProgressStats._totalCarryOverSegmentsToBeDeleted);
+    // Calculate elapsed time based on start of rebalance (global)
+    newOverallProgressStats._estimatedTimeToCompleteAddsInSeconds =
+        calculateEstimatedTimeToCompleteChange(rebalanceProgressStats.getStartTimeMs(),
+            newOverallProgressStats._totalSegmentsToBeAdded, newOverallProgressStats._totalRemainingSegmentsToBeAdded);
+    newOverallProgressStats._estimatedTimeToCompleteDeletesInSeconds =
+        calculateEstimatedTimeToCompleteChange(rebalanceProgressStats.getStartTimeMs(),
+            newOverallProgressStats._totalSegmentsToBeDeleted,
+            newOverallProgressStats._totalRemainingSegmentsToBeDeleted);
+    newOverallProgressStats._averageSegmentSizeInBytes = overallProgressStats._averageSegmentSizeInBytes;
+    newOverallProgressStats._totalEstimatedDataToBeMovedInBytes =
+        overallProgressStats._totalEstimatedDataToBeMovedInBytes
+            + (numAdditionalSegmentsAdded * overallProgressStats._averageSegmentSizeInBytes);
+    newOverallProgressStats._startTimeMs = rebalanceProgressStats.getStartTimeMs();
+
+    return newOverallProgressStats;
+  }
+
+  /**
+   * Calculates the progress stats for the given step or for the overall based on the trigger type
+   * @return the calculated step or progress stats
+   */
+  @VisibleForTesting
+  static TableRebalanceProgressStats.RebalanceProgressStats calculateOverallProgressStats(
```

Comment:
I didn't figure out why need to pass in a `rebalanceProgressStats` to this calculate method?

Reply:
I use `rebalanceProgressStats` for the following:

- For the `IDEAL_STATE_CHANGE_TRIGGER` and `EXTERNAL_VIEW_TO_IDEAL_STATE_CONVERGENCE_TRIGGER`, I get the current overall or step-level stats to do the calculations based on the previous values.
- I also fetch the rebalance `startTimeMs` for the estimated time to complete calculations